s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java

```diff
@@ -233,6 +254,9 @@ public void onSuccess(SslContext sslContext) {
                           SharedResourcePool.forResource(GrpcUtil.SHARED_CHANNEL_EXECUTOR))
                       .newHandler(grpcHandler);
 
+              ctx.pipeline().addAfter(ctx.name(), /* name= */ null,
```

Comment:
Since you are using the same name to add after for both this and the tls handler, the second one should actually go before this.  You might want to give it a name and then add the other after that name to correct that and make it clear what is desired.

Reply:
The intention is to have the order of the handlers be: s2a handler, tls handler, s2a cleanup handler. In https://github.com/grpc/grpc-java/pull/11600/commits/a7fd45f77f1cb93baa6a47d6e845e8cc0fecdd40, I gave names to the tls and cleanup handler; I added the tlsHandler after first, then added the cleanupHandler after the tlsHandler.